cmd/immuadmin/command/hot_backup.go

```diff
@@ -0,0 +1,661 @@
+)
+
+const (
+	prefix            = "IMMUBACKUP"
+	latestFileVersion = 1
+)
+
+const (
+	prefixOffset     = iota
+	versionOffset    = prefixOffset + len(prefix)
+	txIdOffset       = versionOffset + 4
+	txSignSizeOffset = txIdOffset + 8
+	txSizeOffset     = txSignSizeOffset + 4
+	headerSize       = txSizeOffset + 4
+)
+
+var ErrMalformedFile = errors.New("malformed backup file")
+var ErrTxWrongOrder = errors.New("incorrect transaction order in file")
+var ErrTxNotInFile = errors.New("last known transaction not in file")
+
+type commandlineHotBck struct {
+	commandline
+}
+
+func newCommandlineHotBck(os immuos.OS) (*commandlineHotBck, error) {
+	cl := commandline{}
+	cl.config.Name = "immuadmin"
+	cl.context = context.Background()
+	cl.os = os
+
+	return &commandlineHotBck{cl}, nil
+}
+
+func (clb *commandlineHotBck) Register(rootCmd *cobra.Command) *cobra.Command {
+	clb.hotBackup(rootCmd)
+	clb.hotRestore(rootCmd)
+	return rootCmd
+}
+
+type backupParams struct {
+	output   string
+	startTx  uint64
+	append   bool
+	progress bool
+}
+
+func (cl *commandlineHotBck) hotBackup(cmd *cobra.Command) {
+	ccmd := &cobra.Command{
+		Use:   "hot-backup <db_name>",
+		Short: "Make a copy of the database without stopping",
+		Long: "Backup a database to file/stream without stopping the database engine. " +
+			"Backup can run from the beginning or starting from arbitrary transaction.",
+		PersistentPreRunE: cl.ConfigChain(cl.connect),
+		PersistentPostRun: cl.disconnect,
+		RunE: func(cmd *cobra.Command, args []string) error {
+			file := io.Writer(os.Stdout)
+			params, err := prepareBackupParams(cmd.Flags())
+			if err != nil {
+				return err
+			}
+
+			udr, err := cl.immuClient.UseDatabase(cl.context, &schema.Database{DatabaseName: args[0]})
+			if err != nil {
+				return err
+			}
+			cl.context = metadata.NewOutgoingContext(cl.context, metadata.Pairs("authorization", udr.GetToken()))
+
+			if params.output != "-" {
+				f, err := cl.verifyBackupFile(params)
+				if err != nil {
+					return err
+				}
+				defer f.Close()
+				file = f
+			}
+
+			return cl.runHotBackup(file, params.startTx, params.progress)
+		},
+		Args: cobra.ExactArgs(1),
+	}
+	ccmd.Flags().StringP("output", "o", "-", "output file, \"-\" for stdout")
+	ccmd.Flags().Uint64("start-tx", 1, "Transaction ID to start from")
+	ccmd.Flags().Bool("progress", false, "show progress indicator")
+	ccmd.Flags().BoolP("append", "i", false, "append to file, if it already exists (for file output only)")
+	cmd.AddCommand(ccmd)
+}
+
+func prepareBackupParams(flags *pflag.FlagSet) (*backupParams, error) {
+	var params backupParams
+	var err error
+
+	params.output, err = flags.GetString("output")
+	if err != nil {
+		return nil, err
+	}
+	params.append, err = flags.GetBool("append")
+	if err != nil {
+		return nil, err
+	}
+	params.startTx, err = flags.GetUint64("start-tx")
+	if err != nil {
+		return nil, err
+	}
+	params.progress, err = flags.GetBool("progress")
```

Comment:
--progress-bar ?

Reply:
`progress` is just shorter, and I believe understandable.
Do we have anywhere else similar type of options? In `vcn`/`cas` progress is on by default, unless user specifies `--silent` or `-o json`/`-o yaml`

Reply:
curl uses --progress-bar. seeing 'progress' in the source code is not immediate to know if it refers to a progress bar or to some progress measure done during the process

Reply:
changed